spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/DirectedGraph.scala

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.intel.analytics.bigdl.utils
+
+import com.intel.analytics.bigdl.Module
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+/**
+ * Provides useful graph operations for a directed graph.
+ *
+ * The node connection in the graph has direction. This class stores a source node. Note that it
+ * doesn't maintain the topology. The topology of the graph is stored in the connection of the
+ * nodes.
+ * @param source source node of the directed graph
+ * @param reverse use the original direction or the reversed direction
+ * @tparam T Node element type
+ */
+class DirectedGraph[T](source : Node[T], reverse : Boolean = false) {
+
+  /**
+   * Topology sort.
+   * @return A sequence of sorted graph nodes
+   */
+  def topologySort : Array[Node[T]] = {
```

Comment:
I check the output of the DFS based topology sort. Notice that it may loss the parallel property. If we use the indegree based topology sort, we can know which nodes can be run in parallel. This property is useless now as we run a single model per thread but I think it's a useful property to have.

This remind me I should build a separated execution plan for forward and backward.